components/vision.py

```diff
@@ -1,36 +1,91 @@
-            return None
-        else:
-            return self.target_tape_error
-
-    def get_ground_tape_x(self):
-        if not -1 <= self.ground_tape_x <= 1:
-            return None
-        else:
-            return self.ground_tape_x
-
-    def get_ground_tape_y(self):
-        if not -1 <= self.ground_tape_y <= 1:
-            return None
-        else:
-            return self.ground_tape_y
-
-    def get_ground_tape_angle(self):
-        if not -1 <= self.ground_tape_angle <= 1:
-            return None
-        else:
-            return self.ground_tape_angle
+    def __init__(self):
+        self.latency = 0.0
+        self.last_pong = time.time()
+        self.odometry = OrderedDict()
+
+    def execute(self):
+        """Store the current odometry in the queue. Allows projection of target into current position."""
+        self.odometry[time.time()] = (
+            chassis.odometry_x,
+            chassis.odometry_y,
+            chassis.last_heading,
+        )
+        self.ping()
+        self.pong()
+
+    @property
+    def fiducial_in_sight(self):
+        return time.time() - self.fiducial_time - self.latency < 0.5
+
+    def get_fiducial_position(self):
+        """Return the position of the retroreflective fiducials relative to the current robot pose."""
+        vision_time = self.fiducial_time - self.latency
+        if self.fiducial_in_sight:  # Only return if not too stale
+            vision_delta_x, vision_delta_y, vision_delta_heading = _get_pose(
+                vision_time
+            )
+            x = fiducial_x - vision_delta_x
+            y = fiducial_y - vision_delta_y
+            return x, y
+        return None, None
+
+    def _get_pose_delta(self, t):
+        """Search the stored odometry and return the position difference between now and the specified time."""
+        current = next(iter(self.odometry.values()))
+        previous = next(iter(self.odometry.values()))
+        for tm, pose in self.odometry.items():
+            if tm > t:
+                previous = pose
+            else:
+                break
+        # Remove old entries from the back of the queue
+        pops = 0
+        now = time.time()
+        while next(iter(self.odometry.reversed().keys())) < now - 3.0:
+            pops += 1
+        while pops > 0:
+            self.odometry.popitem(last=True)
+            pops -= 1
+        x = current[0] - previous[0]
+        y = current[1] - previous[1]
+        # Rotate to the robot frame of reference
+        # Use the previous heading - that's where we were when the picture was taken
+        heading = previous[2]
+        robot_x = x * math.cos(heading) + y * math.sin(heading)
+        robot_y = -y * math.sin(heading) + y * math.cos(heading)
+        return robot_x, robot_y, current[2] - heading
+
+    def ping(self):
+        """Send a ping to the RasPi to determine the connection latency."""
+        self.ping_time = time.time()
```

Comment:
All you will be measuring here is the amount of time it takes for NetworkTables to decide to send updates to clients, unless you both explicitly flush NetworkTables and ensure that the NetworkTables auto-update hasn't fired within the last 10ms.

Reply:
I think the plan is to slow down the vision code's NT auto-update such that it never actually fires.